_posts/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1.md

```diff
@@ -0,0 +1,117 @@
+
+## Recovery
+
+When operators restart from a checkpoint (automatically during recovery or manually during deployment from a savepoint), the operators first restore their state from the checkpoint storage before resuming the event stream processing.
+
+<span>
+	<center>
+	<img vspace="8" style="width:100%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-4.png" />
+	</center>
+</span>
+
+Since sources are bound to the offsets persisted in the checkpoint, recovery time is often calculated as the sum of the time of the recovery process — outlined in the previous figure — and any additional time needed to process any remaining data up to the point right before the system failure. When an application experiences backpressure, recovery time can also include the total time from the very start of the recovery process until backpressure is fully eliminated.
+
+## Consistency Guarantees
+
+The alignment phase is only necessary for checkpoints with __exactly-once__ processing semantics, which is the default setting in Flink. If an application runs with __at-least-once__ processing semantics, checkpoints will not block any channels with barriers during alignment, which has an additional cost from the duplication of the then-not-blocked events when recovering the operator.
+
+This is not to be confused with having at-least-once semantics only in the sinks — something that many Flink users choose over transactional sinks — because many sink operations are idempotent or converge to the same result (like inputs/outputs to key/value stores). Having at-least-once semantics in an intermediate operator state is often not idempotent (for example a simple count aggregation) and hence using exactly-once checkpoints is advisable for the majority of Flink users.
+
+## Conclusion
+
+This blog post recaps how Flink’s fault tolerance mechanism (based on aligned checkpoints) works, and why checkpointing is a great mechanism for a fault-tolerant stream processor. The checkpointing mechanism has been optimized over time to make checkpoints faster and cheaper (with both asynchronous and incremental checkpoints) and faster-to-recover (local caching), but the basic concepts (barriers, alignment, operator state snapshots) are still the same as in the original version.
+
+The next part will dig into a major break with the original mechanism, namely circumventing the alignment phase — hence the name unaligned checkpoints. Stay tuned for the second part, which explains how unaligned checkpoints work and how they guarantee consistent checkpoint times under backpressure.  
```

Comment:
```suggestion
The next part will dig into a major break with the original mechanism that avoids the alignment phase — the recently-introduced "unaligned checkpoints". Stay tuned for the second part, which explains how unaligned checkpoints work and how they guarantee consistent checkpointing times under backpressure.  
```

Reply:
Waaaaay too complex vocabulary. Remember that not everyone reading this will have the same level of english proficiency.